src/Statistics.cpp

```diff
@@ -6,199 +6,139 @@
-	, m_Depends(a_Depends)
-{}
-
-
-
-
-
-const AString & cStatInfo::GetName(const eStatistic a_Type)
-{
-	ASSERT((a_Type > statInvalid) && (a_Type < statCount));
 
-	return ms_Info[a_Type].m_Name;
+
+cStatManager::cStatManager()
+{
+	for (auto i = allStatisticsNames.begin(); i != allStatisticsNames.end(); i++)
+	{
+		m_MainStats.insert(std::pair<std::string, int>(*i, 0));
+	}
 }
 
 
 
 
 
-eStatistic cStatInfo::GetType(const AString & a_Name)
+StatValue cStatManager::GetValue(const eStatistic &a_Stat) const
 {
-	for (unsigned int i = 0; i < ARRAYCOUNT(ms_Info); ++i)
-	{
-		if (NoCaseCompare(ms_Info[i].m_Name, a_Name) == 0)
-		{
-			return ms_Info[i].m_Type;
-		}
-	}
-
-	return statInvalid;
+	ASSERT((a_Stat > statInvalid) && (a_Stat < statCount));
+	return m_MainStats.at(allStatisticsNames[a_Stat]);
 }
 
 
 
 
 
-eStatistic cStatInfo::GetPrerequisite(const eStatistic a_Type)
+StatValue cStatManager::GetValue(const AString &a_Stat) const
 {
-	ASSERT((a_Type > statInvalid) && (a_Type < statCount));
+	return m_MainStats.at(a_Stat);
+}
 
-	return ms_Info[a_Type].m_Depends;
+
+
+
+
+void cStatManager::SetValue(const eStatistic &a_Stat, const StatValue a_Value)
+{
+	ASSERT((a_Stat > statInvalid) && (a_Stat < statCount));
+	m_MainStats[allStatisticsNames[a_Stat]] = a_Value;
 }
 
 
 
 
 
-cStatManager::cStatManager()
+void cStatManager::SetValue(const AString &a_Stat, const StatValue a_Value)
 {
-	Reset();
+	ASSERT(std::find(allStatisticsNames.begin(), allStatisticsNames.end(), a_Stat) != allStatisticsNames.end());
+	m_MainStats[a_Stat] = a_Value;
 }
 
 
 
 
 
-StatValue cStatManager::GetValue(const eStatistic a_Stat) const
+StatValue cStatManager::AddValue(const eStatistic &a_Stat, const StatValue a_Delta)
 {
 	ASSERT((a_Stat > statInvalid) && (a_Stat < statCount));
 
-	return m_MainStats[a_Stat];
+	m_MainStats[allStatisticsNames[a_Stat]] += a_Delta;
+	return m_MainStats[allStatisticsNames[a_Stat]];
 }
 
 
 
 
 
-void cStatManager::SetValue(const eStatistic a_Stat, const StatValue a_Value)
+StatValue cStatManager::AddValue(const AString &a_Stat, const StatValue a_Delta)
 {
-	ASSERT((a_Stat > statInvalid) && (a_Stat < statCount));
+	ASSERT(std::find(allStatisticsNames.begin(), allStatisticsNames.end(), a_Stat) != allStatisticsNames.end());
+	m_MainStats[a_Stat] += a_Delta;
+	return m_MainStats[a_Stat];
+}
 
-	m_MainStats[a_Stat] = a_Value;
+
+
+
+
+void cStatManager::Reset(void)
+{
+	for (unsigned int i = 0; i < allStatisticsNames.size(); ++i)
+	{
+		m_MainStats[allStatisticsNames[i]] = 0;
+	}
 }
 
 
 
 
 
-StatValue cStatManager::AddValue(const eStatistic a_Stat, const StatValue a_Delta)
+const AString & cStatManager::GetName(const eStatistic a_Type)
 {
-	ASSERT((a_Stat > statInvalid) && (a_Stat < statCount));
+	return allStatisticsNames[a_Type];
+}
 
-	m_MainStats[a_Stat] += a_Delta;
 
-	return m_MainStats[a_Stat];
+
+
+
+const AString & cStatManager::GetPrerequisite(const AString & a_Type)
+{
+	return v1_8AchievementPrerequisites.at(a_Type);
```

Comment:
Protocol-specifics (v1_8) shouldn't bleed into the rest of the server

Reply:
this is simply bad naming from me - any version supported 1.8 - 1.11 uses this table because there weren't any changes

Reply:
Prerequisites sound like they shouldn't be protocol-specific anyway. So this function should return an `enum class Achievement` and get the prerequisite data from the latest 1.16

Reply:
Achievements were remved in 1.12. and replaced with advancements. Up to 1.12 the Achievements were handled like stats in the network

I put this in case mojang added/rermoved something

Reply:
renamed the map and used switch/case